From fc3afac9e59ef200178f7d470fc0b7960d9be064 Mon Sep 17 00:00:00 2001
From: Robin Gong <yibin.gong@nxp.com>
Date: Mon, 1 Feb 2016 11:24:58 +0800
Subject: [PATCH 0911/5242] MLK-12350: ARM: anatop: disable PU regulator on
 i.mx6qp before suspend

commit  0cfbed52b93368159d791cd516d5c158b4d2ef64 from
https://source.codeaurora.org/external/imx/linux-imx.git

To eliminate the power number, need turn off PU regulator before suspend
since it's turned on always on i.mx6qp.

Signed-off-by: Robin Gong <yibin.gong@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/mach-imx/anatop.c |   25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/arch/arm/mach-imx/anatop.c b/arch/arm/mach-imx/anatop.c
index 5f5851a..24945fa 100644
--- a/arch/arm/mach-imx/anatop.c
+++ b/arch/arm/mach-imx/anatop.c
@@ -41,6 +41,8 @@
 #define BM_ANADIG_REG_2P5_ENABLE_WEAK_LINREG	0x40000
 #define BM_ANADIG_REG_2P5_ENABLE_PULLDOWN	0x8
 #define BM_ANADIG_REG_CORE_FET_ODRIVE		0x20000000
+#define BM_ANADIG_REG_CORE_REG1			(0x1f << 9)
+#define BM_ANADIG_REG_CORE_REG2			(0x1f << 18)
 #define BM_ANADIG_ANA_MISC0_STOP_MODE_CONFIG	0x1000
 #define BM_ANADIG_ANA_MISC0_V2_STOP_MODE_CONFIG	0x800
 #define BM_ANADIG_ANA_MISC0_V3_STOP_MODE_CONFIG	0xc00
@@ -96,6 +98,23 @@ static inline void imx_anatop_disconnect_high_snvs(bool enable)
 			BM_ANADIG_ANA_MISC0_DISCON_HIGH_SNVS);
 }
 
+static void imx_anatop_disable_pu(bool off)
+{
+	u32  val, soc;
+	if (off) {
+		regmap_read(anatop, ANADIG_REG_CORE, &val);
+		val &= ~BM_ANADIG_REG_CORE_REG1;
+		regmap_write(anatop, ANADIG_REG_CORE, val);
+	} else {
+		/* track vddpu with vddsoc */
+		regmap_read(anatop, ANADIG_REG_CORE, &val);
+		soc = val & BM_ANADIG_REG_CORE_REG2;
+		val &= ~BM_ANADIG_REG_CORE_REG1;
+		val |= soc >> 9;
+		regmap_write(anatop, ANADIG_REG_CORE, val);
+	}
+}
+
 void imx_anatop_pre_suspend(void)
 {
 	if (cpu_is_imx7d()) {
@@ -109,6 +128,9 @@ void imx_anatop_pre_suspend(void)
 		return;
 	}
 
+	if (cpu_is_imx6q() && imx_get_soc_revision() == IMX_CHIP_REVISION_2_0)
+		imx_anatop_disable_pu(true);
+
 	if ((imx_mmdc_get_ddr_type() == IMX_DDR_TYPE_LPDDR2) &&
 		!imx_gpc_usb_wakeup_enabled())
 		imx_anatop_enable_2p5_pulldown(true);
@@ -134,6 +156,9 @@ void imx_anatop_post_resume(void)
 		return;
 	}
 
+	if (cpu_is_imx6q() && imx_get_soc_revision() == IMX_CHIP_REVISION_2_0)
+		imx_anatop_disable_pu(false);
+
 	if ((imx_mmdc_get_ddr_type() == IMX_DDR_TYPE_LPDDR2) &&
 		!imx_gpc_usb_wakeup_enabled())
 		imx_anatop_enable_2p5_pulldown(false);
-- 
1.7.9.5

